tensorly/contrib/mps_decomposition_cross.py

```diff
@@ -0,0 +1,430 @@
+    d = tl.ndim(input_tensor)
+    fibers_list = []
+
+    # Extract fibers
+    for i in range(rank[k - 1]):
+        for j in range(rank[k]):
+            fiber = row_idx[k - 1][i] + (slice(None, None, None),) + col_idx[k - 1][j]
+            fibers_list.append(fiber)
+
+    if k == d:  # Is[k] will be empty
+        idx = tuple(zip(*row_idx[k - 1])) + (slice(None, None, None),)
+    else:
+        idx = [[] for i in range(d)]
+        for lidx in row_idx[k - 1]:
+            for ridx in col_idx[k - 1]:
+                for j, jj in enumerate(lidx): idx[j].append(jj)
+                for j, jj in enumerate(ridx): idx[len(lidx) + 1 + j].append(jj)
+        idx[k - 1] = slice(None, None, None)
+        idx = tuple(idx)
+
+    core = input_tensor[idx]
+    # shape the core as a 3-d cube
+    core = tl.reshape(core, (rank[k - 1], rank[k], n[k - 1]))
+    core = tl.transpose(core, (0, 2, 1))
+    # merge n_{k-1} and r_k, get a matrix
+    core = tl.reshape(core, (rank[k - 1], n[k - 1] * rank[k]))
+    core = tl.transpose(core)
+
+    # Compute QR decomposition
+    (Q, R) = tl.qr(core)
+    # Maxvol
+    (J, Q_inv) = maxvol(Q)
+    Q_inv = tl.tensor(Q_inv)
+    Q_skeleton = tl.dot(Q, Q_inv)
+
+    # Retrive indices in folded tensor
+    new_idx = [np.unravel_index(idx, [n[k - 1], rank[k]]) for idx in J]  # First retrive idx in folded core
+    next_col_idx = [(jc[0],) + col_idx[k - 1][jc[1]] for jc in new_idx]  # Then reconstruct the idx in the tensor
+
+    return (next_col_idx, fibers_list, Q_skeleton)
+
+
+def maxvol(A):
+    """ Find the rxr submatrix of maximal volume in A(nxr), n>=r
+
+            We want to decompose matrix A as
+                    A = A[:,J] * (A[I,J])^-1 * A[I,:]
+            This algorithm helps us find this submatrix A[I,J] from A, which has the largest determinant.
+            We greedily find vector of max norm, and subtract its projection from the rest of rows.
+
+    Parameters
+    ----------
+
+    A: matrix
+            The matrix to find maximal volume
+
+    Returns
+    -------
+    row_idx: list of int
+            is the list or rows of A forming the matrix with maximal volume,
+    A_inv: matrix
+            is the inverse of the matrix with maximal volume.
+
+    References
+    ----------
+    S. A. Goreinov, I. V. Oseledets, D. V. Savostyanov, E. E. Tyrtyshnikov, N. L. Zamarashkin.
+    How to find a good submatrix.Goreinov, S. A., et al.
+    Matrix Methods: Theory, Algorithms and Applications: Dedicated to the Memory of Gene Golub. 2010. 247-256.
+
+    Ali Çivril, Malik Magdon-Ismail
+    On selecting a maximum volume sub-matrix of a matrix and related problems
+    Theoretical Computer Science. Volume 410, Issues 47–49, 6 November 2009, Pages 4801-4811
+    """
+
+    (n, r) = tl.shape(A)
+
+    # The index of row of the submatrix
+    row_idx = tl.zeros(r)
+
+    # Rest of rows / unselected rows
+    rest_of_rows = tl.tensor(list(range(n)))
+    rest_of_rows = tl.int(rest_of_rows)
```

Comment:
Hello Jean, np.int32 type cannot be used as indices for pytorch, which requires long or byte dtype. Therefore I create int() to coordinate the dtype for indices.

Reply:
Let's just use np.int_ (or int64) then --it should be the same as long--?
However, this raises a good point that we need a way to deal with dtype in a backend agnostic way..

Reply:
It seems np.int64 cannot be used as indices for pytorch either. It expects torch.int64. I don't know how to do it agnostically of backends...

Reply:
Finally, to deal with this problem, I modify the tl.tensor() in pytorch so that it can take dtype = string or numpy.dtype. Now we can use dtype to do indexing instead of int().
Thanks!